src/transforms/key_value_parser.rs

```diff
@@ -0,0 +1,409 @@
+}
+
+#[typetag::serde(name = "key_value_parser")]
+impl TransformConfig for KeyValueConfig {
+    fn build(&self, _cx: TransformContext) -> crate::Result<Box<dyn Transform>> {
+        let field = self
+            .field
+            .as_ref()
+            .unwrap_or(&event::log_schema().message_key());
+
+        let conversions = parse_conversion_map(&self.types)?;
+
+        let trim_key = if let Some(key) = &self.trim_key {
+            Some(key.chars().collect())
+        } else {
+            None
+        };
+
+        let trim_value = if let Some(val) = &self.trim_value {
+            Some(val.chars().collect())
+        } else {
+            None
+        };
+
+        Ok(Box::new(KeyValue {
+            conversions,
+            drop_field: self.drop_field,
+            field: field.clone(),
+            field_split: Option::from(self.field_split.clone()),
+            overwrite_target: self.overwrite_target,
+            separator: self.separator.clone(),
+            target_field: self.target_field.clone(),
+            trim_key,
+            trim_value,
+        }))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn transform_type(&self) -> &'static str {
+        "key_value_parser"
+    }
+}
+
+pub struct KeyValue {
+    conversions: HashMap<Atom, Conversion>,
+    drop_field: bool,
+    field: Atom,
+    field_split: Option<String>,
+    overwrite_target: bool,
+    separator: String,
+    target_field: Option<Atom>,
+    trim_key: Option<Vec<char>>,
+    trim_value: Option<Vec<char>>,
+}
+
+impl KeyValue {
+    pub fn parse_pair(&self, pair: &str) -> Option<(Atom, String)> {
+        let pair = pair.trim();
+        let field_split = self
+            .field_split
+            .as_ref()
+            .ok_or_else(|| Some("".to_string()))
+            .unwrap();
+
+        let fields = if field_split.is_empty() {
+            let mut kv_pair = pair.split_whitespace();
+            let key = kv_pair.next()?;
+            let val = kv_pair.next()?;
+            if kv_pair.next().is_some() {
+                error!(
+                    message = "KV parser saw more than one separator",
+                    rate_limit_secs = 30
+                );
+                return None;
+            }
+
+            (key, val)
+        } else {
+            let split_index = pair.find(field_split).unwrap_or(0);
+            let (key, _val) = pair.split_at(split_index);
+            let key = key.trim();
+            let val = pair[split_index + field_split.len()..].trim();
+
+            if key.is_empty() || val.is_empty() {
+                return None;
+            }
+
+            (key, val)
+        };
+
+        let key = if let Some(trim_key) = &self.trim_key {
+            fields.0.trim_matches(trim_key as &[_])
+        } else {
+            fields.0
+        };
```

Comment:
While this can be done with `map` and `unwrap_or`, I think a `match` is best here:
```suggestion
        let key = match &self.trim_key {
            Some(trim_key) => fields.0.trim_matches(trim_key as &[_]),
            None => fields.0,
        };
```